R/zchunk_L244.building_det.R

```diff
@@ -0,0 +1,1040 @@
+      mutate(variable = if_else(thermal.building.service.input %in% heating_services, "HDD", "CDD")) %>%
+      # Add HDD and CDD
+      left_join_error_no_match(L244.HDDCDD_scen_R_Y, by = c("region", "SRES", "GCM", "variable", "year")) %>%
+      mutate(degree.days = round(value, energy.DIGITS_HDDCDD))
+
+    # L244.GenericServiceSatiation: Satiation levels assumed for non-thermal building services
+    # First, calculate the service output per unit floorspace in the USA region
+    L244.ServiceSatiation_USA <- L144.base_service_EJ_serv %>%
+      filter(GCAM_region_ID == gcam.USA_CODE) %>%
+      # Using left_join_keep_first_only b/c there are repeats and we only need to keep generic columns
+      left_join_keep_first_only(calibrated_techs_bld_det, by = c("sector", "service")) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID") %>%
+      select(region, sector, service, gcam.consumer, nodeInput, building.node.input, building.service.input = supplysector, year, value)
+
+    # Floorspace should be matched in for a specified year, from the full floorspace table (i.e. not one that is subsetted to model base years)
+    L144.flsp_bm2_R_res_Yh <- L144.flsp_bm2_R_res_Yh %>%
+      mutate(gcam.consumer = A44.gcam_consumer$gcam.consumer[grepl("res", A44.gcam_consumer$gcam.consumer)]) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID")
+
+    L144.flsp_bm2_R_comm_Yh <- L144.flsp_bm2_R_comm_Yh %>%
+      mutate(gcam.consumer = A44.gcam_consumer$gcam.consumer[grepl("com", A44.gcam_consumer$gcam.consumer)]) %>%
+      left_join_error_no_match(GCAM_region_names, by = "GCAM_region_ID")
+
+    L244.flsp_bm2_R <- bind_rows(L144.flsp_bm2_R_res_Yh, L144.flsp_bm2_R_comm_Yh) %>%
+      # Again, used to be energy.SATIATION_YEAR, changed to pass timeshift test
+      filter(year == max(BASE_YEARS)) %>%
+      select(-year)
+
+    L244.ServiceSatiation_USA <- L244.ServiceSatiation_USA %>%
+      left_join_error_no_match(L244.flsp_bm2_R %>%
+                                 rename(floorspace_bm2 = value), by = c("region", "gcam.consumer")) %>%
+      left_join_error_no_match(A44.demand_satiation_mult, by = c("building.service.input" = "supplysector")) %>%
+      group_by(region, sector, service) %>%
+      mutate(satiation.level = round(value[year == max(HISTORICAL_YEARS)] * multiplier / floorspace_bm2, energy.DIGITS_CALOUTPUT)) %>%
+      ungroup() %>%
+      select(-multiplier)
+
+    # Generic services: read these values to all regions because they're all the same
+    L244.GenericServiceSatiation <- L244.ServiceSatiation_USA %>%
+      filter(building.service.input %in% generic_services) %>%
+      select(-region, -GCAM_region_ID, -year, -value) %>%
+      distinct() %>%
+      write_to_all_regions(LEVEL2_DATA_NAMES[["GenericServiceSatiation"]], GCAM_region_names = GCAM_region_names)
+
+    # This is bad. Need to match in the floorspace into the base service table, divide to calculate the service demand
```

Comment:
"This is bad."? Does this mean that the calculation is complicated? Or wrong? Maybe clean up the comment and open issues as appropriate.

Reply:
The text of that comment doesn't need to be copied! What I meant was, having to jump through all of these hoops is bad. These are adjustments to the input data that are required in order to cover for functional forms in the model that are not capable of handling heterogeneity, and would cause the model to crash out.

Reply:
All of the "this is bad" comments are from the old data system (along with some more colorful language that I deleted). I can delete them. Here is what Page said about something similar "basically what's going on here is we have to jump through a bunch of hoops on the data side to make up for inflexible functional forms in the model. That said, developing new equations describing the buildings sector and re-coding the thing is probably not going to be a priority in the next few years. Dunno if it's worth an issue on github, Jira, or neither."